### ISSUE TITLE
slave to your milestones

### ISSUE THREAD

**Issue by monkeypants** [bugflow/gh-ticket-gnome#17]
if this file exists `.bugflow/milestone_slave.yaml`, then it should contain something like
```yaml
master_milestone_org: bugflow
master_milestone_repo: auto-gnome
follow_autosprint: true
```
And the auto-gnome must have access to that repo through the GitHub API, otherwise there is an error.
 
If none of the above errors, When a master repo milestone changes, apply the corresponding change to this repo:
 * For every milestone on the master repo, if there is not a milestone with the same name in this repo, create it (and copy details across)
 * If a master repo milestone changes, the same change should be applied to the milestone with the corresponding name in this repo. This includes milestone changes (dates, title, description) and closing a milestone. If the master milestone is deleted, the following milestones are closed instead of deleted.

If the master repo also has `.bugflow/autosprint.yaml` configured (see #???, automatic project board per milestone), and the milestone_slave `follow_autosprint` config is true, then, for every ticket in the slave milestone, if it is not on the master's sprint/milestone project board, add it to the master's sprint/milestone project board (backlog column).

**Comment by monkeypants** [comment]
Note that per #17, we also need to mirror labels (tags) from the master. And if the master is doing autosprint, then the slaves will need to inherit the behavior of moving unfinished tickets to the sorting hat and tagging them with the "did not finish" label.

The milestone mirroring should be idempotent actually, so it shouldn't matter if there is event based and inherited "autosprint" milestone closing behaviors - if there were both, it wouldn't matter what order they were applied in. Which ever tries to happens first will succeed, and the second one will be a no-op.

**Comment by ltankey** [comment]
I think this needs to be rejigged, such that events triggered on the master result in policies running for the master. To invert our flow here, we should have some kind of `propagate_milestones.yaml` policy responsible for ensuring that, for the defined list of slaves/children, milestone changes are propagated out (create, modify, destroy etc.)

```
propagate_milestones.yaml
---
slaves:
  - org/foo_repo
  - org/bar_repo
  - org/baz_repo
```

Permissions are assumed to exist for access to these repos.

This is required because events are triggered on the master here (changes to master milestones), and without a database tracking master/slaves, we won't know where to go and look for config, or to know where we should sync these changes.

**Comment by monkeypants** [comment]
OK, you are right - if the master knew it's slaves, we could avoid a database and do everything just-in-time.

We need both kind of references though:
 * duplicate milestones master needs to know it's slaves, so that it can keep them in sync when the master milestone changes
 * putting own milestone tickets onto the master kanban/project board requires the slave knowing it's master.

**Comment by ltankey** [comment]
Agreed. This will come down to the config required for each policy.

**Comment by monkeypants** [comment]
Right: it would be possible to be a slave to milestones in repository X, but post your tickets to kanban board on repository Y...

**Comment by ltankey** [comment]
Yep! Some new policy that:

```
for ticket in callback:
  if ticket.milestone.name in [b.name for b in config.repo_or_org.boards]:
    sync_ticket_with_board()
```
